src/workerd/api/web-socket.h

```diff
@@ -455,6 +482,22 @@ class WebSocket: public EventTarget {
       // If we are "releasing", we may prevent the websocket from doing certain things like calling
       // send/close. We're more restrictive if we're delivering an Error than delivering a Close.
       HibernatableReleaseState releaseState = HibernatableReleaseState::NONE;
+
+    // There are two possible states for tagsRef:
+    //  1. kj::Function
+    //          - A function pointer that lets us call HibernatableWebSocket::getTags()
+    //            from api::WebSocket.
+    //          - So long as the HibernatableWebSocket is around, we can just call its method.
+    //  2. kj::Array<kj::String>
+    //          - An array of the tags.
+    //          - We're going to be dispatching a Close or an Error event, i.e. the
+    //            HibernatableWebSocket is free to go away. We can no longer rely on its getTags()
+    //            method, so instead we move the data into the api::WebSocket.
+    //
+    // We could just copy all tags into api::WebSocket everytime we reactivate/wake from
+    // hibernation, but it could add up to 2.56KB of memory for each websocket.
+    // With a maximum of 32k websockets, that could put a lot of memory pressure on the DO.
+      kj::OneOf<kj::Function<kj::Array<kj::StringPtr>()>, kj::Array<kj::String>> tagsRef;
```

Comment:
I'm not a fan of having this back door into the HibernatableWebSocket, but it could be that I don't understand the lifetimes of the HibertnabableWebSocket versus the api::WebSocket.

What prevents us from moving the tags into the api::WebSocket?

Reply:
`HibernatableWebSocket`s survive "hibernation" because they live outside of the isolate, `api::WebSocket`s are allocated on the JS heap so they're dropped when we hibernate (synonymous with evict due to inactivity). The primary task of the `HibernatableWebSocket` is to reconstruct the `api::WebSocket` when it is needed again.

There is an awkward phase near the end of the connection when the websocket receives a `CLOSE` (or needs to dispatch an `ERROR` event). The `HibernatableWebSocket` is free to go away after _getOrRecreating_ the `api::WebSocket` and scheduling this final event to run. This happens for a few reasons:
- If we are in the process of dispatching our final `close/error` event, there's no longer any need to keep the `HibernatableWebSocket` since no more events will be dispatched.
- We don't have a way of knowing when the `api::WebSocket` is destroyed. It might end being kept around for a long time (especially if the DO suddenly gets pinned to memory due to a `setInterval` or regular websocket connections).
- There is a hard limit of ~32k `HibernatableWebSocket`s per-actor, so we need to clean them up when we can.

---
We already work around this awkward lifetime issue at the end of the connection<sup>1</sup> by transferring the `kj::Own<kj::WebSocket>` from the `HibernatableWebSocket` to the `api::WebSocket`. Said another way, the `api::WebSocket` doesn't actually own its underlying `kj::WebSocket` (which is used for all the actual websocket-y stuff), it just has a reference to the one owned by the `HibernatableWebSocket`. When we dispatch the final event, we transfer ownership to the `api::WebSocket`. This was implemented here: https://github.com/cloudflare/workerd/pull/664.

> What prevents us from moving the tags into the api::WebSocket?

**Moving**: We only store one copy of each unique tag, since many WebSockets may share it we can't move it.
**Copying**: Mostly just being conscious of memory (for customers) as the comment in that file says. We already cut into the amount of memory available for actors by having them occasionally share isolates, so if we can avoid moving duplicate strings into the isolate by passing `StringPtr`s til the end of the connection then I think that's preferable.

1. I actually forgot about this mechanism, I think we can hitch a ride on the same code for transferring the `kj::WebSocket` and make the lifetimes a lot clearer.

Reply:
> I actually forgot about this mechanism, I think we can hitch a ride on the same code for transferring the kj::WebSocket and make the lifetimes a lot clearer.

Looks like a productive diff! Hopefully this should make things a bit clearer. I'll take another look in the morning and see if I should add any comments.
https://github.com/cloudflare/workerd/compare/4e3a29a2be63f6537d66a2f73a5826cad05de67c..a6af878e38f99db29cc7619ba4585f23bce9813e